File: content/snippets/scala2/interview_lcci/01/06_compress_string.scala
object Lcci0106CompressString {
  def compressString(s: String): String = {
    if (s.isEmpty) {
      return s
    }

    val b = new StringBuilder
    var run = 1

    for (i <- 1 to s.length) {
      if (i < s.length && s(i) == s(i - 1)) {
        run += 1
      } else {
        b.append(s(i - 1))
        b.append(run)
        run = 1
      }
    }

    val compressed = b.toString()
    if (compressed.length < s.length) compressed else s
  }

  def main(args: Array[String]): Unit = {
    println(s"aabcccccaaa -> ${compressString("aabcccccaaa")}")
    println(s"abbccd -> ${compressString("abbccd")}")
  }
}

File: content/snippets/typescript/interview_lcci/01/06_compress_string.ts
function compressString(s: string): string {
  if (s.length === 0) {
    return s;
  }

  let out = "";
  let run = 1;

  for (let i = 1; i <= s.length; i++) {
    if (i < s.length && s[i] === s[i - 1]) {
      run += 1;
      continue;
    }

    out += s[i - 1] + String(run);
    run = 1;
  }

  return out.length < s.length ? out : s;
}

function main(): void {
  console.log(`aabcccccaaa -> ${compressString("aabcccccaaa")}`);
  console.log(`abbccd -> ${compressString("abbccd")}`);
}

main();
